### PR TITLE
[batch] rewrite schema validation so that schema description is working validation function

### PR DESCRIPTION
**Pull request by catoverdrive**
I thought about porting some of the typechecker stuff from hail, but they don't serve quite the same purpose, and the functions themselves are pretty simple. This does move the schema definitions to the bottom of the file rather than at the top, but this way we have a schema description that includes requiredness and is guaranteed to be the thing we're actually validating on.

### CODE REVIEW

File: batch/batch/front_end/validate.py [hunk 1]
```diff
@@ -3,76 +3,138 @@
...
+    def __getitem__(self, key):
+        return self.checkers[key]
+
+    def validate(self, name, obj):
+        super().validate(name, obj)
+        for k in obj:
+            if k not in self.checkers:
+                raise ValidationError(f'unknown key in {name}: {k}')
+        for k, checker in self.checkers.items():
+            if isinstance(checker, required):
```
Comment (danking) [inline batch/batch/front_end/validate.py:47]: The syntax looks a bit weird:
```python3
keyed({
    'foo': required(int),
    'bar': nullabe(int),
    'baz': int
})
```

What do you think of this instead:
```python3
keyed({
    required('foo'): int,
    'bar': nullabe(int),
    'baz': int
})
```
That puts the required ness notion on the right part (the key).

It also simplifies `validate` a bit:
```python3
    def validate(self, name, obj):
         super().validate(name, obj)
         for k in obj:
             if k not in self.checkers:
                 raise ValidationError(f'unknown key in {name}: {k}')
         for k, checker in self.checkers.items():
             if isinstance(k, required) and k not in obj:
                 raise ValidationError(f'{name} missing required key {k}.')
             checker.validate(f"{name}.{key}", obj[k])

```

File: batch/batch/front_end/validate.py [hunk 2]
```diff
@@ -3,76 +3,138 @@
...
+        self.checkers = keyed_checkers
+
+    def __getitem__(self, key):
+        return self.checkers[key]
+
+    def validate(self, name, obj):
+        super().validate(name, obj)
+        for k in obj:
+            if k not in self.checkers:
+                raise ValidationError(f'unknown key in {name}: {k}')
```
Comment (danking) [inline batch/batch/front_end/validate.py:45]: Mild preference to announce all missing keys:
```python3
unknown_keys = obj.keys() - self.checkers.keys()
if unknown_keys:
    raise ValidationError(f'unknown keys in {name}: {", ".join(unknown_keys)}')
```

**Review by danking — CHANGES_REQUESTED** [review]
neat, a few changes. You might consider moving all the validator classes to `hailtop.validate`. That keeps front_end/validate.py short.

**Comment by catoverdrive** [comment]
Oooh, yes. hailtop seems like a good place for it because then we can use it more generally for json schemas, I'll do that.

File: batch/batch/front_end/validate.py (hunk 1 above)
Comment (catoverdrive) [reply batch/batch/front_end/validate.py:47]: yeah, that looks good! I'll change

**Comment by danking** [comment]
needs rebase

**Review by danking — APPROVED** [review]

**Pull request merged**